torch/utils/data/_utils/index_utils.py

```diff
@@ -0,0 +1,150 @@
+from math import sqrt, floor
+
+
+class Range:
+    size: int
+
+    # for slicing
+    start: int
+    stop: int
+    step: int
+
+    # for iterating
+    index: int
+
+    def __init__(self, size: int, start: int = None, stop: int = None, step: int = None):
+        self.size = size
+        self.start = start if start is not None else 0
+        self.stop = stop if stop is not None else size
+        self.step = step if step is not None else 1
+        if not (0 <= self.start <= self.stop):
+            raise ValueError('not (0 <= start <= stop)')
+        if self.stop > self.size:
+            pass  # Allow wrap-around
```

Comment:
I believe stop needs to set as self.size to get rid of the behavior of "wrap-around", in case that users may iterate over dataset multiple times.